### ISSUE TITLE
[clang-tidy] `portability-template-virtual-member-function` warns about virtual method invoked via reference to base class

### ISSUE THREAD

**Issue by tearfur** [llvm/llvm-project#206767]
I need guidance on what is the expected way to fix this warning.

https://godbolt.org/z/813shrrP7

```
<source>:8:10: warning: unspecified virtual member function instantiation; the virtual member function is not instantiated but it might be with a different compiler [portability-template-virtual-member-function]
    8 |     void foo() override {}
      |          ^
<source>:16:16: note: template instantiated here
   16 |     Impl<char> v;
      |                ^
1 warning generated.
```

Imagine this code was in a header included by multiple translation units. Some translation units constructs `Wrapper` objects, some don't. Translation units that construct `Wrapper` won't have this warning, while others will.

All translation units compile, so the there is no concern of the project compiling on GCC/clang but not on MSVC (this was the rationale of the check according to https://github.com/llvm/llvm-project/issues/153155#issuecomment-3243348741).

Now I'm stuck with warnings in translation units that don't construct `Wrapper` objects.

Current trunk: bd6adfedc776c07caf158e59367d9c246c933510

**Comment by vbvictor** [comment]
CC @isuckatcs for visibility

**Comment by carlosgalvezp** [comment]
The check essentially bans virtual member functions in class templates, because its unspecified behavior whether they get instantiated or not (as per the linked comment), which can lead to portability issues. 

Similar to having a check banning `x = f() + g();` because the order of evaluation is unspecified.

IIUC your options are remove the virtual functions or disable the check.

**Comment by isuckatcs** [comment]
It looks like since the check was introduced, it has been causing more confusion than good. I'll think about an RFC to revert it.

**Comment by vbvictor** [comment]
I think one improvement to this check could be a `note:` diagnostic attached to main warning advising to delete this function. So users will have clear picture: to satisfy the check they either have to remove the function or remove the check.

**Comment by tearfur** [comment]
I decided to disable the check because I don't see a way of fixing it.

**Comment by isuckatcs** [comment]
@tearfur WDYT about the reworded diag message in #211122? Is it more helpful than the current one?

**Comment by tearfur** [comment]
It definitely explains the rationale of the check better.

But I still hope something can be done (doesn't have to be in #211122) to address the case I raised in this ticket, where we know the class template will instantiate just fine if any translation unit instantiates it, but the check flags it anyway.

**Comment by isuckatcs** [comment]
> But I still hope something can be done [...] to address the case [...] where we know the class template will instantiate just fine [...]

For this we need to reason about the body of the class template, including the bodies of the member functions for every possible type. Even if clang-tidy was capable of doing this, there are still infinite amount of types the class can be instantiated with, and reasoning about those is not possible.

**Comment by tearfur** [comment]
That's too bad. Thanks for following up on this.

Feel free to go ahead with the PR if everybody else has a consensus.
